docs/source/format/CDeviceDataInterface.rst

```diff
@@ -650,6 +652,331 @@ The stream source is not assumed to be thread-safe. Consumers wanting to
+                       int code, const char* message, const char* metadata);
+
+      // release callback
+      void (*release)(struct ArrowAsyncDeviceStreamHandler* self);
+
+      // opaque handler-specific data
+      void* private_data;
+    };
+
+    #endif  // ARROW_C_ASYNC_STREAM_INTERFACE
+
+.. note::
+    The canonical guard ``ARROW_C_ASYNC_STREAM_INTERFACE`` is meant to avoid
+    duplicate definitions if two projects copy the C async stream interface
+    definitions into their own headers, and a third-party project includes
+    from these two projects. It is therefore important that this guard is kept
+    exactly as-is when these definitions are copied.
+
+The ArrowAsyncDeviceStreamHandler structure
+'''''''''''''''''''''''''''''''''''''''''''
+
+The structure has the following fields:
+
+.. c:member:: int (*ArrowAsyncDeviceStreamHandler.on_schema)(struct ArrowAsyncDeviceStreamHandler*, struct ArrowAsyncProducer*, struct ArrowSchema*, const char*)
+
+    *Mandatory.* Handler for receiving the schema of the stream. All records should
+    match the provided schema. If successful, the function should return 0, otherwise
+    it should return an ``errno``-compatible error code.
+
+    The ``const char*`` parameter exists for producers to provide any extra contextual information
+    they want, such as the total number of rows in the stream, statistics, or otherwise. This is
+    encoded in the same format as :c:member:`ArrowSchema.metadata`. If not ``NULL``,
+    the lifetime is only the scope of the call to this function. A consumer who wants to maintain
+    the additional metadata beyond the lifetime of this call *MUST* copy the value themselves.
+
+    Unless the ``on_error`` handler is called, this will always get called exactly once and will be
+    the first method called on this object. As such the producer *MUST* provide an ``ArrowAsyncProducer``
+    object when calling this function to allow the consumer to manage back-pressure and flow control.
+    The producer maintains ownership of the ``ArrowAsyncProducer`` and must clean it up before or after
+    calling the release callback on this object.
+
+    A producer that receives a non-zero result here must not subsequently call anything other than
+    the release callback on this object.
+
+.. c:member:: int (*ArrowAsyncDeviceStreamHandler.on_next_task)(struct ArrowAsyncDeviceStreamHandler*, struct ArrowAsyncTask*, const char*)
+
+    *Mandatory.* Handler to be called when a new record is available for processing. The
+    schema for each record should be the same as the schema that ``on_schema`` was called with.
+    If successfully handled, the function should return 0, otherwise it should return an
+    ``errno``-compatible error code.
+
+    Rather than passing the record itself it receives an ``ArrowAsyncTask`` instead to facilitate
+    better consumer-focused thread control as far as receiving the data. A call to this function
```

Comment:
Why does it facilitate enough in that regard? This only seems to complicate things: the data is available already, why not give it already to the consumer?

Reply:
Ok, I've read the explanation below.